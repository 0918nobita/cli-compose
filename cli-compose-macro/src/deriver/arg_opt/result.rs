use derive_more::Display;
use proc_macro2::TokenStream;

#[derive(Debug, Display)]
pub enum ArgOptErrKind {
    #[display(
        fmt = "#[derive(ArgOpt)] can only be applied to structs with single unnamed field or enums"
    )]
    InvalidTypeDef,
}

pub struct ArgOptErr {
    kind: ArgOptErrKind,
    tokens: TokenStream,
}

impl ArgOptErr {
    pub fn new(kind: ArgOptErrKind, tokens: TokenStream) -> Self {
        Self { kind, tokens }
    }
}

impl From<ArgOptErr> for syn::Error {
    fn from(err: ArgOptErr) -> syn::Error {
        syn::Error::new_spanned(err.tokens, err.kind.to_string())
    }
}
