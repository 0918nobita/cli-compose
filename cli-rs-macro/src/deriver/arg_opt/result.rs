use derive_more::Display;
use proc_macro2::TokenStream;

#[derive(Debug, Display)]
pub enum ArgOptErrKind {
    #[display(
        fmt = "#[derive(ArgOpt)] can only be applied to structs with single unnamed field or enums"
    )]
    InvalidTypeDef,

    #[display(fmt = "Literals in #[arg_opt(..)] is not allowed")]
    UnexpectedLit,

    #[display(fmt = "Metadata in #[arg_opt(..)] is invalid")]
    InvalidMeta,

    #[display(fmt = "#[arg_opt(long = ..)] must be a string literal")]
    InvalidLongValue,

    #[display(fmt = "#[arg_opt(short = ..)] must be a char literal")]
    InvalidShortValue,

    #[display(fmt = "Unexpected key in #[arg_opt(..)]")]
    UnexpectedKey,
}

pub struct ArgOptErr {
    kind: ArgOptErrKind,
    tokens: TokenStream,
}

impl ArgOptErr {
    pub fn new(kind: ArgOptErrKind, tokens: TokenStream) -> Self {
        Self { kind, tokens }
    }
}

impl From<ArgOptErr> for syn::Error {
    fn from(err: ArgOptErr) -> syn::Error {
        syn::Error::new_spanned(err.tokens, err.kind.to_string())
    }
}

pub type ArgOptResult<T> = Result<T, ArgOptErr>;
