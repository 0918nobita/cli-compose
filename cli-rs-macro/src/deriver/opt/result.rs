use derive_more::Display;
use proc_macro2::TokenStream;

#[derive(Debug, Display)]
pub enum OptErrKind {
    #[display(fmt = "#[derive(Opt)] can only be applied to empty structs")]
    InvalidStruct,

    #[display(fmt = "Literals in #[opt(..)] are not allowed")]
    UnexpectedLit,

    #[display(fmt = "Metadata in #[opt(..)] is invalid")]
    InvalidMeta,

    #[display(fmt = "#[opt(long = ..)] must be a string literal")]
    InvalidLongValue,

    #[display(fmt = "#[opt(short = ..)] must be a char literal")]
    InvalidShortValue,

    #[display(fmt = "Unexpected key in #[opt(..)]")]
    UnexpectedKey,
}

pub struct OptErr {
    kind: OptErrKind,
    tokens: TokenStream,
}

impl OptErr {
    pub fn new(kind: OptErrKind, tokens: TokenStream) -> Self {
        Self { kind, tokens }
    }
}

impl From<OptErr> for syn::Error {
    fn from(err: OptErr) -> Self {
        syn::Error::new_spanned(err.tokens, err.kind.to_string())
    }
}

pub type OptResult<T> = Result<T, OptErr>;
