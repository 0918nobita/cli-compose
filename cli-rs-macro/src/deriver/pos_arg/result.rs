use derive_more::Display;
use proc_macro2::TokenStream;

#[derive(Debug, Display)]
pub enum PosArgErrKind {
    #[display(fmt = "#[derive(PosArg)] can only be applied to structs with single unnamed field")]
    InvalidStruct,
}

pub struct PosArgErr {
    kind: PosArgErrKind,
    tokens: TokenStream,
}

impl PosArgErr {
    pub fn new(kind: PosArgErrKind, tokens: TokenStream) -> Self {
        Self { kind, tokens }
    }
}

impl From<PosArgErr> for syn::Error {
    fn from(err: PosArgErr) -> Self {
        syn::Error::new_spanned(err.tokens, err.kind.to_string())
    }
}
