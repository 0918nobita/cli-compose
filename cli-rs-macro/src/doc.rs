use syn::{Attribute, Lit, Meta};

fn try_get_single_line_doc(attr: &Attribute) -> Option<String> {
    let meta = attr.parse_meta().ok()?;

    let lit_str = match meta {
        Meta::NameValue(syn::MetaNameValue {
            path,
            lit: Lit::Str(lit_str),
            ..
        }) if path.is_ident("doc") => lit_str,
        _ => return None,
    };

    Some(lit_str.value().trim_start().to_owned())
}

pub fn extract_doc(attrs: &[Attribute]) -> String {
    attrs
        .iter()
        .filter_map(try_get_single_line_doc)
        .collect::<Vec<_>>()
        .join("\n")
}
