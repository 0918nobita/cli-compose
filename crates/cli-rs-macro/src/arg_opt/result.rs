use derive_more::Display;
use proc_macro2::TokenStream;

#[derive(Debug, Display)]
pub enum ArgOptErrorKind {
    #[display(
        fmt = "#[derive(ArgOpt)] can only be applied to structs with single unnamed field or enums"
    )]
    InvalidTypeDef,

    #[display(fmt = "Literals in #[arg_opt(..)] is not allowed")]
    UnexpectedLit,

    #[display(fmt = "Metadata in #[arg_opt(..)] is invalid")]
    InvalidMeta,

    #[display(fmt = "#[arg_opt(long = ..)] must be a string literal")]
    InvalidLongValue,

    #[display(fmt = "#[arg_opt(short = ..)] must be a char literal")]
    InvalidShortValue,

    #[display(fmt = "Unexpected key in #[arg_opt(..)]")]
    UnexpectedKey,
}

pub struct ArgOptError {
    kind: ArgOptErrorKind,
    tokens: TokenStream,
}

impl ArgOptError {
    pub fn new(kind: ArgOptErrorKind, tokens: TokenStream) -> Self {
        Self { kind, tokens }
    }
}

impl From<ArgOptError> for syn::Error {
    fn from(err: ArgOptError) -> syn::Error {
        syn::Error::new_spanned(err.tokens, err.kind.to_string())
    }
}

pub type ArgOptResult<T> = Result<T, ArgOptError>;
