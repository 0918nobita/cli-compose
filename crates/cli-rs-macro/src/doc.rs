use syn::Attribute;

fn try_get_single_line_doc(attr: Attribute) -> Option<String> {
    let meta = attr.parse_meta().ok()?;
    match meta {
        syn::Meta::NameValue(syn::MetaNameValue {
            path,
            lit: syn::Lit::Str(lit_str),
            ..
        }) if path.is_ident("doc") => Some(lit_str.value().trim_start().to_owned()),
        _ => None,
    }
}

pub fn extract_doc(attrs: impl Iterator<Item = Attribute>) -> String {
    attrs
        .filter_map(try_get_single_line_doc)
        .collect::<Vec<_>>()
        .join("\n")
}
