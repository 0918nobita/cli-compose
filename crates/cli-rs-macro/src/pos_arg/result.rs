use derive_more::Display;
use proc_macro2::TokenStream;

#[derive(Debug, Display)]
pub enum PosArgErrKind {
    #[display(fmt = "#[derive(PosArg)] can only be applied to structs with single unnamed field")]
    InvalidStruct,

    #[display(fmt = "Literals in #[pos_arg(..)] are not allowed")]
    UnexpectedLit,

    #[display(fmt = "Metadata in #[pos_arg(..)] is invalid")]
    InvalidMeta,

    #[display(fmt = "#[pos_arg(name = ..)] must be a string literal")]
    InvalidNameValue,

    #[display(fmt = "Unexpected key in #[pos_arg(..)]")]
    UnexpectedKey,
}

pub struct PosArgErr {
    kind: PosArgErrKind,
    tokens: TokenStream,
}

impl PosArgErr {
    pub fn new(kind: PosArgErrKind, tokens: TokenStream) -> Self {
        Self { kind, tokens }
    }
}

impl From<PosArgErr> for syn::Error {
    fn from(err: PosArgErr) -> Self {
        syn::Error::new_spanned(err.tokens, err.kind.to_string())
    }
}

pub type PosArgResult<T> = Result<T, PosArgErr>;
