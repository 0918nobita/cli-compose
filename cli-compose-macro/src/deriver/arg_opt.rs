mod result;

use convert_case::{Case, Casing};
use darling::{
    ast::{self, Data},
    FromDeriveInput,
};
use proc_macro2::TokenStream;
use quote::{quote, ToTokens};

use self::result::{ArgOptErr, ArgOptErrKind};
use crate::doc::extract_doc;

#[derive(FromDeriveInput)]
#[darling(attributes(arg_opt), forward_attrs(doc))]
struct ArgOptInput {
    ident: syn::Ident,

    data: Data<syn::Ident, syn::Field>,

    attrs: Vec<syn::Attribute>,

    #[darling(default)]
    long: Option<String>,

    #[darling(default)]
    short: Option<char>,

    #[darling(default)]
    use_default: bool,
}

pub fn derive_arg_opt(input: TokenStream) -> syn::Result<TokenStream> {
    let input = syn::parse2::<syn::DeriveInput>(input)?;

    let input = match ArgOptInput::from_derive_input(&input) {
        Ok(input) => input,
        Err(err) => return Ok(err.write_errors()),
    };

    match &input.data {
        Data::Enum(_) => {
            let enum_name = &input.ident;

            let long = input
                .long
                .unwrap_or_else(|| input.ident.to_string().to_case(Case::Kebab));

            let flag = match input.short {
                Some(short) => {
                    quote! { cli_compose::Flag::BothLongAndShort(#long.to_owned(), #short) }
                }
                None => quote! { cli_compose::Flag::LongOnly(#long.to_owned()) },
            };

            let doc = extract_doc(&input.attrs);

            Ok(quote! {
                impl cli_compose::AsArgOpt for #enum_name {
                    fn flag() -> cli_compose::Flag {
                        #flag
                    }

                    fn description() -> String {
                        #doc.to_owned()
                    }

                    fn parse(s: &str) -> Option<Self> {
                        <#enum_name as std::str::FromStr>::from_str(s).ok()
                    }
                }
            })
        }

        Data::Struct(
            fields @ ast::Fields {
                style: ast::Style::Tuple,
                fields: fields_vec,
                ..
            },
        ) => {
            let field = match &fields_vec[..] {
                [field] => field,
                _ => {
                    return Err(ArgOptErr::new(
                        ArgOptErrKind::InvalidTypeDef,
                        fields.to_token_stream(),
                    )
                    .into())
                }
            };

            let struct_name = &input.ident;
            let long = input
                .long
                .unwrap_or_else(|| struct_name.to_string().to_case(Case::Kebab));

            let flag = match input.short {
                Some(short) => {
                    quote! { cli_compose::Flag::BothLongAndShort(#long.to_owned(), #short) }
                }
                None => quote! { cli_compose::Flag::LongOnly(#long.to_owned()) },
            };

            let doc = extract_doc(&input.attrs);

            let ty = field.ty.clone();

            Ok(quote! {
                impl cli_compose::AsArgOpt for #struct_name {
                    fn flag() -> cli_compose::Flag {
                        #flag
                    }

                    fn description() -> String {
                        #doc.to_owned()
                    }

                    fn parse(s: &str) -> Option<Self> {
                        let val = <#ty as std::str::FromStr>::from_str(s).ok()?;
                        Some(#struct_name(val))
                    }
                }
            })
        }

        Data::Struct(fields) => {
            Err(ArgOptErr::new(ArgOptErrKind::InvalidTypeDef, fields.to_token_stream()).into())
        }
    }
}
