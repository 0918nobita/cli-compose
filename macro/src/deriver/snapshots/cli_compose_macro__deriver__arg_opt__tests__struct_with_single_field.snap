---
source: cli-compose-macro/src/deriver/arg_opt.rs
assertion_line: 159
expression: "test_derive_arg_opt(quote! { struct Foo(String) ; }).unwrap()"
---
impl cli_compose::schema::AsArgOpt for Foo {
    fn flag() -> cli_compose::schema::Flag {
        cli_compose::schema::Flag::LongOnly("foo".to_owned())
    }
    fn description() -> String {
        "".to_owned()
    }
    fn parse(s: &str) -> Option<Self> {
        let val = <String as std::str::FromStr>::from_str(s).ok()?;
        Some(Foo(val))
    }
}

